[Google for Woo] Hub Menu functionality part 1

hafizrahman opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of: #11822 

Closes #11949
<!-- Id number of the GitHub issue this PR addresses. -->

Please do not merge until `fluxCVersion` points to `trunk-` something.

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This plugin does these:
- Rename `IsGoogleListingsEnabled` to `IsGoogleForWooEnabled` because they're gonna rename the plugin that way soon.
- Updates the requirement in `IsGoogleForWooEnabled` to check for plugin version, too.
- Updates `GoogleRepository` to add fetching for Google Ads campaign.
- Adds `HasGoogleAdsCampaigns` use case
- Add testing logic when tapping Google for WooCommerce button in hub menu, to make testing this PR easy:
     - When opened for first time, it should open to the campaign creation screen
     - When opened again, it should open  campaign dashboard screen

### Not handled in this PR yet
As this is just a part 1, these things are known to be missing/inefficient so far:
1. The logic for tapping the button is only there for testing purposes to ensure both creation and dashboard URLs work with different test site setups. The correct logic will be added in a future PR.
2. There is no handling / detection for successful campaign creation yet, it will be added in a future PR.
3. When tapping the button, it takes a while before the webview opens, because it runs a fetch in `hasGoogleAdsCampaigns()` each time. I'll refactor this in a future PR.



### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

1. Start with a test site with active GLA plugin, existing active campaigns, and plugin version of 2.7.5 or above.
2. Start app, go to hub menu, ensure the **Google for WooCommerce** button exists
3. Tap the button,
4. Ensure that a webview opens,
5. The webview will ask you to login first if you haven't logged in in the webview before, please do so,
6. After logging in, please ensure you are redirected to the Google Ads dashboard screen

### Optional testing: campaign creation URL
1. Modify line 277 of `MoreMenuViewModel` from  `selectedSite.get().adminUrlOrDefault + AppUrls.GOOGLE_ADMIN_DASHBOARD` to `selectedSite.get().adminUrlOrDefault + AppUrls.GOOGLE_ADMIN_CAMPAIGN_CREATION_SUFFIX`
2. Restart app, go to hub menu, ensure the **Google for WooCommerce** button exists
3. Tap the button,
4. Ensure that a webview opens,
5. The webview will ask you to login first if you haven't logged in in the webview before, please do so,
6. After logging in, please ensure you are redirected to the Google Ads campaign creation screen

### Video
[Screen_recording_20240703_190452.webm](https://github.com/woocommerce/woocommerce-android/assets/266376/11212670-3f58-403f-903a-0ee9533c5e47)

hafizrahman committed:
Update fluxc

hafizrahman committed:
Add repository.

hafizrahman committed:
Update eligibility checking using repository.

Also renames file based on incoming new plugin name.

hafizrahman committed:
Merge branch 'trunk' into feat/11822-gla-hub-menu-functionality

* trunk: (222 commits)
  Update `PlayStoreStrings.po` for 19.2
  Update full_description watch details in all languages
  Improve `create_gh_release` documentation
  Use only `w` suffix on wear app version name
  Remove redundant `wear-` prefix to release name
  Fix detekt indentation issues
  Add toolbar to new product creation with AI fragment
  Rename new fragment
  Move code to new package
  CI: Add yaml language server schema to allow ides to recognize yml files
  CI: Remove custom meta-data to ci schedule for dependency analysis
  CI: Add agent attribute with queue android to every yml file
  Saved order id when created
  Fixed formatting
  Communicate status of the order creation to the totals
  Model for communication order creation status
  Simplified loading state
  Removed delay in producation code
  Fixed background color
  Text on the button as in the design
  ...

hafizrahman committed:
Update usecase name.

dangermattic commented:
<!--
  0 Errors
  2 Warnings: Class <code>GoogleRepository</...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>GoogleRepository</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/225">19.5</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
Add new requirement to check plugin version first.

hafizrahman committed:
Update fluxc

hafizrahman committed:
Update repository with fetch campaigns function

hafizrahman committed:
Add use case for checking if user there are google ads campaigns.

hafizrahman committed:
Add Google button logic.

hafizrahman committed:
Update unit tests

hafizrahman committed:
Merge branch 'trunk' into feat/11822-gla-hub-menu-functionality

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/moremenu/MoreMenuViewModel.kt:
```diff
@@ -272,6 +277,24 @@ class MoreMenuViewModel @Inject constructor(
     private fun onPromoteProductsWithGoogle() {
         WooLog.d(WooLog.T.GOOGLE_ADS, "onPromoteProductsWithGoogle")
+
+        launch {
+            val urlToOpen = when {
+                hasCreatedGoogleAdsCampaign || hasGoogleAdsCampaigns() -> {
+                    selectedSite.get().adminUrlOrDefault + AppUrls.GOOGLE_ADMIN_DASHBOARD
+                }
+                else -> {
+                    selectedSite.get().adminUrlOrDefault + AppUrls.GOOGLE_ADMIN_CAMPAIGN_CREATION_SUFFIX
+                }
+            }
+
+            triggerEvent(MoreMenuEvent.ViewAdminEvent(urlToOpen))
+            // This is just temporary to test this function,
+            // in practice we want to set this to true if a campaign is successfully created in webview.
+            if (!hasCreatedGoogleAdsCampaign) {
+                hasCreatedGoogleAdsCampaign = true
+            }
+        }
```
While testing this part, I noticed that if the user is not logged in in the web view after logging in with their WP.com account, they won't land on the `GOOGLE_ADMIN_DASHBOARD` or the campaign creation flow. They'll land on Calypso generic dashboard: 

<img width="300"  src="https://github.com/woocommerce/woocommerce-android/assets/2663464/bff0f581-4c18-4a79-af0a-52938e5fa275"> 

Might be worth checking that flow. Also it took really long the Webview to log in, and there wasn't any spinner or anything indicating that it was login in. So, I actually thought that it was frozen.

Also minor thing. Testing logic relying on `hasCreatedGoogleAdsCampaign` to open different URLs, won't work when you have campaigns created because `hasGoogleAdsCampaigns()` will be true all the time and thus it would load `GOOGLE_ADMIN_DASHBOARD` always. 
In any case that is irrelevant as it is testing logic :)

JorgeMucientes replied:
Hmm, weird @hafizrahman. In a subsequent test while testing the other PR, I went through the login process again and everything worked as expected. Unsurer what happened in my first test, but very unlikely the app code is related to it. 

https://github.com/woocommerce/woocommerce-android/assets/2663464/f1a85552-f623-4405-8903-54a37b5e14a3

hafizrahman replied:
As we have tested together in Slack, the commit 8ef8dcff28ce9e95ddedd55ffa68933633da2bfb seems to work well to deal with wpcom login.

JorgeMucientes commented:
Nice work @hafizrahman 👍🏼 
I left some feedback about a possible issue when logging in through the webview. Might be worth checking, but I don't think it should block the merging of this PR. So feel free to investigate/address that in a different GH issue. Approving 🙂!

hafizrahman committed:
Try: adding auto auth for wpcom users.

hafizrahman committed:
Detekt fix

hafizrahman committed:
simplify mock

hafizrahman committed:
Update plugin check to use wc/v3/system_status endpoint instead

This allows store manager role to be able to use the feature as well. Previously, the plugin endpoint is admin role only.

hafizrahman committed:
detekt fixes

hafizrahman committed:
Merge branch 'trunk' into feat/11822-gla-hub-menu-functionality

* trunk: (32 commits)
  Fix scanning image spinner spacing
  Rename string res
  Remove unnecessary alpha
  Improve wording replacing "uploaded" with "selected"
  Ignore WireMock and its dependencies in Dependabot configuration
  Add constraints for WireMock transitive dependencies
  Use `dependency-submission@v3`
  add test collector token for tablet run
  fix updated reference
  Update RELEASE-NOTES
  Add monochrome version of ic_launcher to support themed icons
  Use xml version of ic_launcher_foreground for v26
  Add reversed app icon in xml format to use for wasabi variant
  Use ic_launcher_background.xml instead of color/ic_launcher_background
  Add ic_launcher_background.xml
  Reformat ic_launcher_foreground.xml
  Remove the background from ic_launcher_foreground.xml
  Fix detekt issues
  Fix minor issue with error handling
  Error handling when scanning image
  ...
